lib/http-signatures/src/cavage/safety_check.rs

```diff
@@ -0,0 +1,108 @@
+use http::{header::DATE, Method, Request};
+use std::time::{Duration, SystemTime, SystemTimeError};
+use thiserror::Error;
+
+use super::SignatureHeader;
+
+/// 15 minutes
+const MAX_ACCEPTED_SIGNATURE_AGE: Duration = Duration::from_secs(15 * 60);
+
+const REQUIRED_GET_HEADERS: &[&str] = &["host"];
+const REQUIRED_POST_HEADERS: &[&str] = &["host", "content-type", "digest"];
+
+#[derive(Debug, Error)]
+pub enum SafetyCheckError {
+    #[error(transparent)]
+    InvalidDateHeader(#[from] httpdate::Error),
+
+    #[error(transparent)]
+    InvalidHeaderValue(#[from] http::header::ToStrError),
+
+    #[error(transparent)]
+    InvalidSystemTime(#[from] SystemTimeError),
+
+    #[error("Missing required headers")]
+    MissingRequiredHeaders,
+
+    #[error("Signature too old")]
+    SignatureTooOld,
+
+    #[error("Unsupported HTTP method")]
+    UnsupportedHttpMethod,
+}
+
+#[inline]
+fn is_subset<I>(left: &[I], right: &[I]) -> bool
+where
+    I: PartialEq,
+{
+    if left.len() <= right.len() {
+        left.iter().all(|item| right.contains(item))
+    } else {
+        false
+    }
+}
+
+pub fn is_safe<'a, B, I, S>(
+    req: &Request<B>,
+    signature_header: &SignatureHeader<'_, I, S>,
+) -> Result<(), SafetyCheckError>
+where
+    I: Iterator<Item = &'a str> + Clone,
+{
+    let collected_headers = signature_header.headers.clone().collect::<Vec<&str>>();
+    let is_subset = match *req.method() {
+        Method::GET => is_subset(REQUIRED_GET_HEADERS, &collected_headers),
+        Method::POST => is_subset(REQUIRED_POST_HEADERS, &collected_headers),
+        _ => return Err(SafetyCheckError::UnsupportedHttpMethod),
+    };
+
+    if !is_subset {
+        return Err(SafetyCheckError::MissingRequiredHeaders);
+    }
+
+    // Check if the `headers` field either includes `date` or `(created)`
+    if !collected_headers.contains(&"date") && !collected_headers.contains(&"(created)") {
+        return Err(SafetyCheckError::MissingRequiredHeaders);
+    }
+
+    if let Some(created) = signature_header.created {
+        let created_time = SystemTime::UNIX_EPOCH + Duration::from_secs(created);
+        let now = SystemTime::now();
+
+        if now.duration_since(created_time)? > MAX_ACCEPTED_SIGNATURE_AGE {
```

Comment:
What if we have explicit `(expires)` pseudo header? should we follow it or should we `min(MAX_ACCEPTED_SIGNATURE_AGE, expires)`?

Reply:
I'd go for a `min` in that case The smaller the time window, the better

Reply:
Added a check